Feature: add support for text and uuid type control columns

akulapid opened the pull request:
This change introduces two new parameters on create_parent, a pair of encoder/decoder functions that users can define to describe how a text/uuid column maps to time information. Once we derive time from the column, we can leverage existing time partitioning logic to manage partitions for text columns.

Allowing users to define functions brings flexibility to support various formats of identifiers such as UUIDv7, ULID, Snowflake IDs etc. As a convenience, encoder/decoder functions are included for UUIDv7 given the widespread standard.

Issue tracker: https://github.com/pgpartman/pg_partman/issues/528

akulapid committed:
Feature: add support for text and uuid type control columns

This change introduces two new parameters on create_parent, a
pair of encoder/decoder functions that users can define to
describe how a text/uuid column maps to time information. Once
we derive time from the column, we can leverage existing time
partitioning logic to manage partitions for text columns.

Allowing users to define functions brings flexibility to support
various formats of identifiers such as UUIDv7, ULID, Snowflake IDs
etc. As a convenience, encoder/decoder functions are included for
UUIDv7 given the widespread standard.

keithf4 commented on sql/functions/create_partition_time.sql:
```diff
@@ -272,6 +291,7 @@ FOREACH v_time IN ARRAY p_partition_times LOOP
         SELECT
             sub_parent
             , sub_control
+            , sub_time_encoder
```
Doesn't subpartitioning also need the decoder function as well?

akulapid replied:
Good catch! I missed one use case completely to allow sub partitioning with text columns. Have fixed this and also included a test for this

keithf4 commented on sql/functions/show_partition_info.sql:
```diff
@@ -103,10 +110,12 @@ ELSE
 END IF;
 
-IF v_control_type = 'time' OR (v_control_type = 'id' AND v_epoch <> 'none') THEN
+IF v_control_type = 'time' OR (v_control_type = 'id' AND v_epoch <> 'none') OR (v_control_type IN ('text', 'uuid')) THEN
```
Simplify conditional here as well

akulapid replied:
Addressed

keithf4 commented on sql/functions/show_partition_name.sql:
```diff
@@ -64,7 +67,7 @@ partition_schema := v_parent_schema;
 SELECT general_type INTO v_control_type FROM @extschema@.check_control_type(v_parent_schema, v_parent_tablename, v_control);
 
-IF ( (v_control_type = 'time') OR (v_control_type = 'id' AND v_epoch <> 'none') ) THEN
+IF ( (v_control_type = 'time') OR (v_control_type = 'id' AND v_epoch <> 'none') OR (v_control_type IN ('text', 'uuid')) ) THEN
```
Simplify conditional here as well

akulapid replied:
Addressed

keithf4 commented on sql/functions/show_partitions.sql:
```diff
@@ -86,11 +90,10 @@ END IF;
 
-
-IF v_control_type = 'time' THEN
+IF v_control_type = 'time' OR (v_control_type IN ('text', 'uuid')) THEN
```
Simplify conditional

keithf4 commented on sql/functions/show_partitions.sql:
```diff
@@ -86,11 +90,10 @@ END IF;
 
     v_sql := v_sql || format('
-        ORDER BY (regexp_match(pg_get_expr(c.relpartbound, c.oid, true), $REGEX$\(([^)]+)\) TO \(([^)]+)\)$REGEX$))[1]::text::timestamptz %s '
+        ORDER BY (regexp_match(pg_get_expr(c.relpartbound, c.oid, true), $REGEX$\(([^)]+)\) TO \(([^)]+)\)$REGEX$))[1] %s '
```
This has to convert to a timestamp to get the expected logical order out no? Otherwise doesn't this sort by text ordering?

keithf4 replied:
The intention of this function is to show them in the logical, decoded time order. Otherwise the other functions that use this function to get the list of child tables for things like retention may act on the wrong child tables.

akulapid replied:
Understood. I was thinking for text column partitioning, one would always use lexicographically sorted identifiers like uuidv7/ulid, so this query should implicitly be time ordered.
Do you think it's safer to explicitly convert to timestamp and sort?

keithf4 replied:
I'd like to keep it consistently time ordered for all time-based partitioning. So I think it would be safer to convert it here.

akulapid replied:
Done, have pushed a new commit with this change

keithf4 commented on sql/functions/undo_partition.sql:
```diff
@@ -109,7 +115,7 @@ IF v_parent_tablename IS NULL THEN
 
 SELECT general_type INTO v_control_type FROM @extschema@.check_control_type(v_parent_schema, v_parent_tablename, v_control);
-IF v_control_type = 'time' OR (v_control_type = 'id' AND v_epoch <> 'none') THEN
+IF v_control_type = 'time' OR (v_control_type = 'id' AND v_epoch <> 'none') OR (v_control_type IN ('text', 'uuid')) THEN
```
Simplify conditional

akulapid replied:
Addressed

keithf4 commented on sql/functions/undo_partition.sql:
```diff
@@ -277,18 +286,27 @@ LOOP
     <<inner_child_loop>>
     LOOP
-        IF v_control_type = 'time' OR (v_control_type = 'id' AND v_epoch <> 'none') THEN
+        IF v_control_type = 'time' OR (v_control_type = 'id' AND v_epoch <> 'none') OR (v_control_type IN ('text', 'uuid')) THEN
```
Simplify conditional

akulapid replied:
Addressed

keithf4 reviewed (changes requested):
This is an awesome PR! Thank you!

Some questions and syntax corrections to review

akulapid committed:
address review comments:

simplify time based partitioning condition expressions
fix sub-partitioning using text/uuid control type on sub partitions
revert change to fix sort order for time columns in show_partitions()

akulapid commented on sql/functions/create_parent.sql:
```diff
@@ -323,7 +329,7 @@ LOOP
 END LOOP;
 
-IF v_control_type = 'time' OR (v_control_type = 'id' AND p_epoch <> 'none') THEN
+IF v_control_type = 'time' OR (v_control_type = 'id' AND p_epoch <> 'none') OR (v_control_type IN ('text', 'uuid')) THEN
```
Yes, I was trying to keep the change incremental, but this is indeed cleaner. Have addressed this in all the places

akulapid commented on sql/functions/drop_partition_time.sql:
```diff
@@ -111,7 +114,7 @@ END IF;
 SELECT general_type INTO v_control_type FROM @extschema@.check_control_type(v_parent_schema, v_parent_tablename, v_control);
 IF v_control_type <> 'time' THEN
-    IF (v_control_type = 'id' AND v_epoch = 'none') OR v_control_type <> 'id' THEN
+    IF (v_control_type = 'id' AND v_epoch = 'none') OR v_control_type <> 'id' OR (v_control_type IN ('text', 'uuid') AND v_time_encoder IS NULL) THEN
```
Updated the message

akulapid commented:
Hi @keithf4, just wanted to follow up on this, hope this addresses the changes. Let me know if there's anything else needed from my side

keithf4 commented:
> Hi @keithf4, just wanted to follow up on this, hope this addresses the changes. Let me know if there's anything else needed from my side

Yes, I am keeping an eye on it! One additional note on the `show_partitions` function. I may be slow to respond for a bit. Just have a lot of other stuff going on. Shared this PR around the office and people are really impressed with it. So thank you!

akulapid commented:
Good to hear! I understand this may take some time, will try not to bump it up too often :)

akulapid committed:
address review comments:

convert partition bounds to timestamp in show_partitions to ensure
that the result is time-ordered

keithf4 commented:
This looks really good! 

Only have one more ask and it doesn't have to be done as part of this commit either (I may have this merged before you even get to it). Any chance you could write a test with the uuid7 datatype with this? I understand it has another non-default datatype requirement, but since that was one of the main reasons for requesting this feature, I'd like to have a test for it since it may be a common usage going forward.

Make a subfolder in the `test` folder called `test_uuid7` and place them in there. I'm working on organizing the tests so that if specific requirements are needed for them to pass, those tests are all isolated in their own folder.

akulapid committed:
address review comments:

add tests for sub-partitioning with uuid7 column
improve test comments to describe column data types clearly

akulapid commented:
Good point, I missed this scenario. I have the commit ready, I can push to this same branch if you'd like to keep everything together before merging. Otherwise, happy to raise a separate PR. Let me know.

rauanmayemir commented:
Anxious to try this. 😄

keithf4 commented:
> Anxious to try this. 😄

Go ahead and commit to this branch. Sorry I missed your first reply.

akulapid commented:
Done, have pushed the change. There are some conflicts now with development branch though, and I wasn't able to rebase because there are failing tests on development. Is this something you can take up when merging? Otherwise, I can rebase once development is stable again

keithf4 commented:
Hmm didn't realize failing tests would block that. I'll try and get it fixed when I have a chance.

keithf4 commented:
I fixed the merge conflicts for now. 
One of the new uuid tests is failing. I'll try and look into it further when I have a chance, but figured I'd get this fixed so you might be able to look at it too before I am able to.
```
pg_prove -ovf test/test-uuid-time-subpart-custom-start.sql
```
```
...
ok 232 - Check id_taptest_table_p20_p20241101 does not exist
ok 233 - Check id_taptest_table_p20_p20241031 does not exist
not ok 234 - Check id_taptest_table_p30_p20241103 does not exist
# Failed test 234: "Check id_taptest_table_p30_p20241103 does not exist"
not ok 235 - Check id_taptest_table_p30_p20241102 does not exist
# Failed test 235: "Check id_taptest_table_p30_p20241102 does not exist"
not ok 236 - Check id_taptest_table_p30_p20241101 does not exist
# Failed test 236: "Check id_taptest_table_p30_p20241101 does not exist"
not ok 237 - Check id_taptest_table_p30_p20241031 does not exist
# Failed test 237: "Check id_taptest_table_p30_p20241031 does not exist"
not ok 238 - Check id_taptest_table_p40_p20241103 does not exist
# Failed test 238: "Check id_taptest_table_p40_p20241103 does not exist"
not ok 239 - Check id_taptest_table_p40_p20241102 does not exist
# Failed test 239: "Check id_taptest_table_p40_p20241102 does not exist"
not ok 240 - Check id_taptest_table_p40_p20241101 does not exist
# Failed test 240: "Check id_taptest_table_p40_p20241101 does not exist"
not ok 241 - Check id_taptest_table_p40_p20241031 does not exist
# Failed test 241: "Check id_taptest_table_p40_p20241031 does not exist"
not ok 242 - Check id_taptest_table_p50_p20241103 does not exist
# Failed test 242: "Check id_taptest_table_p50_p20241103 does not exist"
not ok 243 - Check id_taptest_table_p50_p20241102 does not exist
# Failed test 243: "Check id_taptest_table_p50_p20241102 does not exist"
ok 244 - Check id_taptest_table_p50_p20241101 does not exist
ok 245 - Check id_taptest_table_p50_p20241031 does not exist
not ok 246 - Check id_taptest_table_p60_p20241103 does not exist
# Failed test 246: "Check id_taptest_table_p60_p20241103 does not exist"
not ok 247 - Check id_taptest_table_p60_p20241102 does not exist
# Failed test 247: "Check id_taptest_table_p60_p20241102 does not exist"
...
```

akulapid commented:
I hope development is not failing only for me? If not, could I suggest holding off this merge until that is fixed?
test-id-time-subpart-custom-start is failing for me on development, so some sub-partitioning logic seems to be broken there. My changes, which were up-to-date with master, should otherwise be green. It should be easier to merge this once development is also green, and we shouldn't really expect any more test failures at that point.
Let me know if this makes sense. I hope I'm not missing something about the workflow here!

keithf4 commented:
I've fixed the issue with `test-id-time-subpart-custom-start` not working. Was a regression in trying to adjust when retention runs. All the tests in the top level of the `test` folder are now passing for me if you want to give it another try

akulapid commented:
Thanks, can confirm it's passing for me too. Just wanted to clarify once on how to proceed next. I generally prefer to rebase PR changes with the trunk, only merging into trunk and not the other way. This way the history is cleaner and the PR only has changes pertaining to it.

Would it be okay if I reverted your merge commits on this branch, instead rebase the original changes with development? I'll need to force push this branch, but the history should be cleaner in the end. Or let me know if I should do it another way.

keithf4 commented:
Sure if that works for you, go ahead. Hopefully the commits I did can help you fix the conflicts that I resolved with them. If not, I'm happy to try and resolve them as well if they're still there.

akulapid commented:
Yep, I'm cross checking the conflicts with this. Have rebased and pushed, hope this should be good now

keithf4 commented on sql/functions/run_maintenance.sql:
```diff
@@ -219,11 +221,21 @@ LOOP
-                            ) INTO v_child_timestamp;
+
+            IF v_control_type = 'time' OR (v_control_type = 'id' AND v_row.epoch <> 'none') THEN
+                EXECUTE format('SELECT %s::text FROM %I.%I'
```
Missing the LIMIT 1 here

akulapid replied:
Ah, can't believe I missed that. Have fixed it now.
Maintenance was still working though, select into seems to be taking the first row in any case

keithf4 replied:
True, but I want to ensure the plan has the LIMIT clause in it for optimization

keithf4 commented on sql/functions/run_maintenance.sql:
```diff
@@ -219,11 +221,21 @@ LOOP
+                                    , v_row_max_time.partition_tablename
+                                ) INTO v_child_timestamp;
+            ELSIF v_control_type IN ('text', 'uuid') THEN
+                EXECUTE format('SELECT %s(%s::text) FROM %I.%I'
```
Missing the LIMIT 1 here

akulapid replied:
Fixed

akulapid committed:
address review comments:

fix upstream conflict in run_maintenance query
